sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -738,6 +739,49 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan] with PredicateHelpe
 }
 
 /**
+ * Reorder associative integral-type operators and fold all constants into one.
+ */
+object ReorderAssociativeOperator extends Rule[LogicalPlan] {
+  private def isAssociativelyFoldable(e: Expression): Boolean =
+    e.deterministic && e.isInstanceOf[BinaryArithmetic] && e.dataType.isInstanceOf[IntegralType] &&
+      isSingleOperatorExpr(e)
+
+  private def isSingleOperatorExpr(e: Expression): Boolean = e.find {
+    case a: Add if a.getClass == e.getClass => false
+    case m: Multiply if m.getClass == e.getClass => false
+    case _: BinaryArithmetic => true
+    case _ => false
+  }.isEmpty
+
+  private def getOperandList(e: Expression): Seq[Expression] = e match {
+    case BinaryArithmetic(a, b) => getOperandList(a) ++ getOperandList(b)
+    case other => other :: Nil
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsDown {
```

Comment:
how about
```
def flattenAdd(e: Expression): Seq[Expression] = e match {
  case Add(l, r) => flattenAdd(l) ++ flattenAdd(r)
  case other => other
}

...
plan transformAllExpressions {
  case a: Add if a.deterministic && a.dataType.isInstanceOf[IntegralType] =>
    val (foldables, others) => flattenAdd(a).partition(_.foldable)
    if (foldables.size > 1) {
      val foldableExpr = foldables.reduce(Add(_, _))
      val c = Literal.create(foldableExpr.eval(), a.dataType)
      if (others.isEmpty) c else Add(others.reduce(Add(_, _)), c)
    } else {
      a
    }
}
```

We can duplicate some code for `Multiply`, and I think this maybe more readable than the current version.